src/org/zaproxy/zap/extension/pscanrulesAlpha/CSPNonceScanner.java

```diff
@@ -0,0 +1,201 @@
+import net.htmlparser.jericho.HTMLElementName;
+import net.htmlparser.jericho.Source;
+
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PassiveScanThread;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+/**
+ * Invalid CSP Nonce passive scan rule
+ *
+ * @author
+ */
+public class CSPNonceScanner extends PluginPassiveScanner{
+
+	private static final String MESSAGE_PREFIX = "pscanalpha.cspnonce.";
+	private static final int PLUGIN_ID = 19999; // TODO: get assigned an ID
+	private static final String CSP_HEADER = "Content-Security-Policy";
+
+	// https://w3c.github.io/webappsec-csp/#grammardef-nonce-source
+	private static final Pattern NONCE_SRC_PATT = Pattern.compile("nonce-([^ ,;]*)");
+	private static final Pattern VALID_NONCE_PATT = Pattern.compile("[A-Za-z0-9+/-_]+=?=?");
+
+	private enum VulnType {CSP_NONCE_REUSED, CSP_INVALID_NONCE};
+
+	private PassiveScanThread parent = null;
+	private static final Logger logger = Logger.getLogger(CSPNonceScanner.class);
+
+	// nonce-srcs seen in responses
+	private HashSet<String> Nonces = new HashSet<String>();
+
+	@Override
+	public void setParent(PassiveScanThread parent) {
+		this.parent = parent;
+	}
+
+	@Override
+	public void scanHttpRequestSend(HttpMessage msg, int id) {
+		// Only checking the response for this plugin
+	}
+
+	private void raiseAlert(VulnType currentVT, String evidence, HttpMessage msg, int id) {
+		Alert alert = new Alert(getPluginId(), //PluginID
+					getRisk(currentVT),
+					Alert.CONFIDENCE_HIGH, //Reliability
+					getAlertElement(currentVT, "name")); //Name
+			alert.setDetail(
+				getAlertElement(currentVT, "desc"), //Description
+				msg.getRequestHeader().getURI().toString(), //URI
+				"",	// Param
+				"", // Attack
+				"", // Other info
+				getAlertElement(currentVT, "soln"), //Solution
+				getAlertElement(currentVT, "refs"), //References
+				evidence,	// Evidence
+					16, // CWE-16: Configuration
+					15,	//WASC-15: Application Misconfiguration
+				msg); //HttpMessage
+			parent.raiseAlert(id, alert);
+	}
+
+	private boolean isNonceSrcValid(String nonceSrc) {
+		Matcher validNonceMatcher = VALID_NONCE_PATT.matcher(nonceSrc);
+		return validNonceMatcher.matches();
+	}
+
+	private List<String> getNonceSrcs(String header) {
+		List<String> nonceSrcs = new ArrayList<String>();
+		Matcher nonceSrcMatcher = NONCE_SRC_PATT.matcher(header);
+
+		while (nonceSrcMatcher.find()) {
+			nonceSrcs.add(nonceSrcMatcher.group(1));
+		}
+		return nonceSrcs;
+	}
+
+	private void checkNonces(HttpMessage msg, int id, String header) {
+		for (String nonceSrc: getNonceSrcs(header)) {
+			if (!isNonceSrcValid(nonceSrc)) {
+				raiseAlert(VulnType.CSP_INVALID_NONCE, nonceSrc, msg, id);
+			}
+			if (Nonces.contains(nonceSrc)) {
+				raiseAlert(VulnType.CSP_NONCE_REUSED, nonceSrc, msg, id);

```

Comment:
I've been thinking about this more. A nonce registry (of sorts) could be added but it would need a few things.

- The nonce values
- An association to the msg it was seen it (HistoryReference perhaps)
- An element like source hostname or something to differentiate when a match is really a match. If a nonce happens to match across completely different hosts or apps would it matter? [Since all proxied messages would be passively scanned]
- A pruning mechanism, so that we don't encounter memory issues perhaps limit the registry to 10k samples per hostname (or whatever the differentiating detail is)...after 10k is there statistic relevant to a match? (20k?)

Food for thought...